From d0bfd4ecf777621b2b9a4e698d7ee9f6f9ce095d Mon Sep 17 00:00:00 2001
From: He Junyan <junyan.he@intel.com>
Date: Sun, 4 Jun 2023 00:10:36 +0800
Subject: [PATCH 101/127] va: Map drm fourcc into va fourcc in video format

The fourcc defined in va.h and drm_fourcc.h sometimes is not identical.
For example, the I420 format is defined as "I420" in va.h but defined
as "YU12" in drm_fourcc.h.
---
 .../gst-libs/gst/va/gstvavideoformat.c        | 173 ++++++++++++------
 .../gst-libs/gst/va/gstvavideoformat.h        |   6 +
 2 files changed, 121 insertions(+), 58 deletions(-)

diff --git a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvavideoformat.c b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvavideoformat.c
index 35c711c659..48f728124e 100644
--- a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvavideoformat.c
+++ b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvavideoformat.c
@@ -23,6 +23,9 @@
 #endif
 
 #include "gstvavideoformat.h"
+#ifndef G_OS_WIN32
+#include <libdrm/drm_fourcc.h>
+#endif
 
 #define GST_CAT_DEFAULT gst_va_display_debug
 GST_DEBUG_CATEGORY_EXTERN (gst_va_display_debug);
@@ -35,62 +38,70 @@ static struct FormatMap
   GstVideoFormat format;
   guint va_rtformat;
   VAImageFormat va_format;
+  /* The drm fourcc may have different definition from VA */
+  guint drm_fourcc;
 } format_map[] = {
-#define F(format, fourcc, rtformat, order, bpp, depth, r, g, b, a) {      \
+#define F(format, fourcc, drm, rtformat, order, bpp, depth, r, g, b, a) {  \
     G_PASTE (GST_VIDEO_FORMAT_, format),                                \
     G_PASTE (VA_RT_FORMAT_, rtformat),                             \
     { VA_FOURCC fourcc, G_PASTE (G_PASTE (VA_, order), _FIRST),    \
-      bpp, depth, r, g, b, a } }
-#define G(format, fourcc, rtformat, order, bpp) \
-    F (format, fourcc, rtformat, order, bpp, 0, 0, 0 ,0, 0)
-  G (NV12, ('N', 'V', '1', '2'), YUV420, NSB, 12),
-  G (NV21, ('N', 'V', '2', '1'), YUV420, NSB, 21),
-  G (VUYA, ('A', 'Y', 'U', 'V'), YUV444, LSB, 32),
-  F (RGBA, ('R', 'G', 'B', 'A'), RGB32, LSB, 32, 32, 0x000000ff,
-      0x0000ff00, 0x00ff0000, 0xff000000),
-  F (RGBx, ('R', 'G', 'B', 'X'), RGB32, LSB, 32, 24, 0x000000ff,
-      0x0000ff00, 0x00ff0000, 0x00000000),
-  F (BGRA, ('B', 'G', 'R', 'A'), RGB32, LSB, 32, 32, 0x00ff0000,
-      0x0000ff00, 0x000000ff, 0xff000000),
-  F (ARGB, ('A', 'R', 'G', 'B'), RGB32, LSB, 32, 32, 0x0000ff00,
-      0x00ff0000, 0xff000000, 0x000000ff),
-  F (xRGB, ('X', 'R', 'G', 'B'), RGB32, LSB, 32, 24, 0x0000ff00,
-      0x00ff0000, 0xff000000, 0x00000000),
-  F (ABGR, ('A', 'B', 'G', 'R'), RGB32, LSB, 32, 32, 0xff000000,
-      0x00ff0000, 0x0000ff00, 0x000000ff),
-  F (xBGR, ('X', 'B', 'G', 'R'), RGB32, LSB, 32, 24, 0xff000000,
-      0x00ff0000, 0x0000ff00, 0x00000000),
-  F (BGRx, ('B', 'G', 'R', 'X'), RGB32, LSB, 32, 24, 0x00ff0000,
-      0x0000ff00, 0x000000ff, 0x00000000),
-  G (UYVY, ('U', 'Y', 'V', 'Y'), YUV422, NSB, 16),
-  G (YUY2, ('Y', 'U', 'Y', '2'), YUV422, NSB, 16),
-  G (AYUV, ('A', 'Y', 'U', 'V'), YUV444, LSB, 32),
+      bpp, depth, r, g, b, a }, drm }
+#ifndef G_OS_WIN32
+#define G(format, fourcc, drm, rtformat, order, bpp) \
+    F (format, fourcc, drm, rtformat, order, bpp, 0, 0, 0 ,0, 0)
+#else
+#define G(format, fourcc, drm, rtformat, order, bpp) \
+    F (format, fourcc, 0, rtformat, order, bpp, 0, 0, 0 ,0, 0)
+#endif
+  G (NV12, ('N', 'V', '1', '2'), DRM_FORMAT_NV12, YUV420, NSB, 12),
+  G (NV21, ('N', 'V', '2', '1'), DRM_FORMAT_NV21, YUV420, NSB, 21),
+  G (VUYA, ('A', 'Y', 'U', 'V'), DRM_FORMAT_AYUV, YUV444, LSB, 32),
+  F (RGBA, ('R', 'G', 'B', 'A'), DRM_FORMAT_RGBA8888, RGB32, LSB, 32, 32,
+      0x000000ff, 0x0000ff00, 0x00ff0000, 0xff000000),
+  F (RGBx, ('R', 'G', 'B', 'X'), DRM_FORMAT_RGBX8888, RGB32, LSB, 32, 24,
+      0x000000ff, 0x0000ff00, 0x00ff0000, 0x00000000),
+  F (BGRA, ('B', 'G', 'R', 'A'), DRM_FORMAT_BGRA8888, RGB32, LSB, 32, 32,
+      0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000),
+  F (ARGB, ('A', 'R', 'G', 'B'), DRM_FORMAT_ARGB8888, RGB32, LSB, 32, 32,
+      0x0000ff00, 0x00ff0000, 0xff000000, 0x000000ff),
+  F (xRGB, ('X', 'R', 'G', 'B'), DRM_FORMAT_XRGB8888, RGB32, LSB, 32, 24,
+      0x0000ff00, 0x00ff0000, 0xff000000, 0x00000000),
+  F (ABGR, ('A', 'B', 'G', 'R'), DRM_FORMAT_ABGR8888, RGB32, LSB, 32, 32,
+      0xff000000, 0x00ff0000, 0x0000ff00, 0x000000ff),
+  F (xBGR, ('X', 'B', 'G', 'R'), DRM_FORMAT_XBGR8888, RGB32, LSB, 32, 24,
+      0xff000000, 0x00ff0000, 0x0000ff00, 0x00000000),
+  F (BGRx, ('B', 'G', 'R', 'X'), DRM_FORMAT_BGRX8888, RGB32, LSB, 32, 24,
+      0x00ff0000, 0x0000ff00, 0x000000ff, 0x00000000),
+  G (UYVY, ('U', 'Y', 'V', 'Y'), DRM_FORMAT_UYVY, YUV422, NSB, 16),
+  G (YUY2, ('Y', 'U', 'Y', '2'), DRM_FORMAT_YUYV, YUV422, NSB, 16),
+  G (AYUV, ('A', 'Y', 'U', 'V'), DRM_FORMAT_AYUV, YUV444, LSB, 32),
   /* F (????, NV11), */
-  G (YV12, ('Y', 'V', '1', '2'), YUV420, NSB, 12),
+  G (YV12, ('Y', 'V', '1', '2'), DRM_FORMAT_YVU420, YUV420, NSB, 12),
   /* F (????, P208), */
-  G (I420, ('I', '4', '2', '0'), YUV420, NSB, 12),
+  G (I420, ('I', '4', '2', '0'), DRM_FORMAT_YUV420, YUV420, NSB, 12),
   /* F (????, YV24), */
   /* F (????, YV32), */
   /* F (????, Y800), */
   /* F (????, IMC3), */
   /* F (????, 411P), */
   /* F (????, 411R), */
-  G (Y42B, ('4', '2', '2', 'H'), YUV422, LSB, 16),
+  G (Y42B, ('4', '2', '2', 'H'), DRM_FORMAT_YUV422, YUV422, LSB, 16),
   /* F (????, 422V), */
   /* F (????, 444P), */
-  G (RGBP, ('R', 'G', 'B', 'P'), RGBP, LSB, 8),
+  /* No RGBP support in drm fourcc */
+  G (RGBP, ('R', 'G', 'B', 'P'), DRM_FORMAT_INVALID, RGBP, LSB, 8),
   /* F (????, BGRP), */
   /* F (????, RGB565), */
   /* F (????, BGR565), */
-  G (Y210, ('Y', '2', '1', '0'), YUV422_10, NSB, 32),
+  G (Y210, ('Y', '2', '1', '0'), DRM_FORMAT_Y210, YUV422_10, NSB, 32),
   /* F (????, Y216), */
-  G (Y410, ('Y', '4', '1', '0'), YUV444_10, NSB, 32),
-  G (Y212_LE, ('Y', '2', '1', '2'), YUV422_12, NSB, 32),
-  G (Y412_LE, ('Y', '4', '1', '2'), YUV444_12, NSB, 32),
+  G (Y410, ('Y', '4', '1', '0'), DRM_FORMAT_Y410, YUV444_10, NSB, 32),
+  G (Y212_LE, ('Y', '2', '1', '2'), DRM_FORMAT_Y212, YUV422_12, NSB, 32),
+  G (Y412_LE, ('Y', '4', '1', '2'), DRM_FORMAT_Y412, YUV444_12, NSB, 32),
   /* F (????, Y416), */
   /* F (????, YV16), */
-  G (P010_10LE, ('P', '0', '1', '0'), YUV420_10, NSB, 24),
-  G (P012_LE, ('P', '0', '1', '2'), YUV420_12, NSB, 24),
+  G (P010_10LE, ('P', '0', '1', '0'), DRM_FORMAT_P010, YUV420_10, NSB, 24),
+  G (P012_LE, ('P', '0', '1', '2'), DRM_FORMAT_P012, YUV420_12, NSB, 24),
   /* F (P016_LE, P016, ????), */
   /* F (????, I010), */
   /* F (????, IYUV), */
@@ -98,19 +109,20 @@ static struct FormatMap
   /* F (????, A2B10G10R10), */
   /* F (????, X2R10G10B10), */
   /* F (????, X2B10G10R10), */
-  G (GRAY8, ('Y', '8', '0', '0'), YUV400, NSB, 8),
-  G (Y444, ('4', '4', '4', 'P'), YUV444, NSB, 24),
+  /* No GRAY8 support in drm fourcc */
+  G (GRAY8, ('Y', '8', '0', '0'), DRM_FORMAT_INVALID, YUV400, NSB, 8),
+  G (Y444, ('4', '4', '4', 'P'), DRM_FORMAT_YUV444, YUV444, NSB, 24),
   /* F (????, Y16), */
   /* G (VYUY, VYUY, YUV422), */
   /* G (YVYU, YVYU, YUV422), */
   /* F (ARGB64, ARGB64, ????), */
   /* F (????, ABGR64), */
-  F (RGB16, ('R', 'G', '1', '6'), RGB16, NSB, 16, 16, 0x0000f800,
-     0x000007e0, 0x0000001f, 0x00000000),
-  F (RGB, ('R', 'G', '2', '4'), RGB32, NSB, 32, 24, 0x00ff0000,
-     0x0000ff00, 0x000000ff, 0x00000000),
-  F (BGR10A2_LE, ('A', 'R', '3', '0'), RGB32, LSB, 32, 30, 0x3ff00000,
-     0x000ffc00, 0x000003ff, 0x30000000),
+  F (RGB16, ('R', 'G', '1', '6'), DRM_FORMAT_RGB565, RGB16, NSB, 16, 16,
+     0x0000f800, 0x000007e0, 0x0000001f, 0x00000000),
+  F (RGB, ('R', 'G', '2', '4'), DRM_FORMAT_RGB888, RGB32, NSB, 32, 24,
+     0x00ff0000, 0x0000ff00, 0x000000ff, 0x00000000),
+  F (BGR10A2_LE, ('A', 'R', '3', '0'), DRM_FORMAT_ARGB2101010, RGB32, LSB, 32, 30,
+     0x3ff00000, 0x000ffc00, 0x000003ff, 0x30000000),
 #undef F
 #undef G
 };
@@ -119,6 +131,7 @@ static const struct RBG32FormatMap
 {
   GstVideoFormat format;
   VAImageFormat va_format[2];
+  guint drm_fourcc;
 } rgb32_format_map[] = {
 #define  F(fourcc, order, bpp, depth, r, g, b, a)                       \
   {  VA_FOURCC fourcc, G_PASTE (G_PASTE (VA_, order), _FIRST), bpp, depth, r, g, b, a }
@@ -127,35 +140,43 @@ static const struct RBG32FormatMap
   { GST_VIDEO_FORMAT_ARGB, {
       A (('B', 'G', 'R', 'A'), LSB, 0x0000ff00, 0x00ff0000, 0xff000000, 0x000000ff),
       A (('A', 'R', 'G', 'B'), MSB, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000),
-    } },
+    },
+    DRM_FORMAT_BGRA8888 },
   { GST_VIDEO_FORMAT_RGBA, {
       A (('A', 'B', 'G', 'R'), LSB, 0x000000ff, 0x0000ff00, 0x00ff0000, 0xff000000),
       A (('R', 'G', 'B', 'A'), MSB, 0xff000000, 0x00ff0000, 0x0000ff00, 0x000000ff),
-    } },
+    },
+    DRM_FORMAT_ABGR8888 },
   { GST_VIDEO_FORMAT_ABGR, {
       A (('R', 'G', 'B', 'A'), LSB, 0xff000000, 0x00ff0000, 0x0000ff00, 0x000000ff),
       A (('A', 'B', 'G', 'R'), MSB, 0x000000ff, 0x0000ff00, 0x00ff0000, 0xff000000),
-    } },
+    },
+    DRM_FORMAT_RGBA8888 },
   { GST_VIDEO_FORMAT_BGRA, {
       A (('A', 'R', 'G', 'B'), LSB, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000),
       A (('B', 'G', 'R', 'A'), MSB, 0x0000ff00, 0x00ff0000, 0xff000000, 0x000000ff),
-    } },
+    },
+    DRM_FORMAT_ARGB8888 },
   { GST_VIDEO_FORMAT_xRGB, {
       X (('B', 'G', 'R', 'X'), LSB, 0x0000ff00, 0x00ff0000, 0xff000000),
       X (('X', 'R', 'G', 'B'), MSB, 0x00ff0000, 0x0000ff00, 0x000000ff),
-    } },
+    },
+    DRM_FORMAT_BGRX8888 },
   { GST_VIDEO_FORMAT_RGBx, {
       X (('X', 'B', 'G', 'R'), LSB, 0x000000ff, 0x0000ff00, 0x00ff0000),
       X (('R', 'G', 'B', 'X'), MSB, 0xff000000, 0x00ff0000, 0x0000ff00),
-    } },
+    },
+    DRM_FORMAT_XBGR8888 },
   { GST_VIDEO_FORMAT_xBGR, {
       X (('R', 'G', 'B', 'X'), LSB, 0xff000000, 0x00ff0000, 0x0000ff00),
       X (('X', 'B', 'G', 'R'), MSB, 0x000000ff, 0x0000ff00, 0x00ff0000),
-    } },
+    },
+    DRM_FORMAT_RGBX8888 },
   { GST_VIDEO_FORMAT_BGRx, {
       X (('X', 'R', 'G', 'B'), LSB, 0x00ff0000, 0x0000ff00, 0x000000ff),
       X (('B', 'G', 'R', 'X'), MSB, 0x0000ff00, 0x00ff0000, 0xff000000),
-    } },
+    },
+    DRM_FORMAT_XRGB8888 },
 #undef X
 #undef A
 #undef F
@@ -175,6 +196,19 @@ get_format_map_from_va_fourcc (guint va_fourcc)
   return NULL;
 }
 
+static const struct FormatMap *
+get_format_map_from_drm_fourcc (guint drm_fourcc)
+{
+  int i;
+
+  for (i = 0; i < G_N_ELEMENTS (format_map); i++) {
+    if (format_map[i].drm_fourcc == drm_fourcc)
+      return &format_map[i];
+  }
+
+  return NULL;
+}
+
 static struct FormatMap *
 get_format_map_from_video_format (GstVideoFormat format)
 {
@@ -244,6 +278,22 @@ gst_va_fourcc_from_video_format (GstVideoFormat format)
   return map ? map->va_format.fourcc : 0;
 }
 
+GstVideoFormat
+gst_va_video_format_from_drm_fourcc (guint fourcc)
+{
+  const struct FormatMap *map = get_format_map_from_drm_fourcc (fourcc);
+
+  return map ? map->format : GST_VIDEO_FORMAT_UNKNOWN;
+}
+
+guint
+gst_va_drm_fourcc_from_video_format (GstVideoFormat format)
+{
+  const struct FormatMap *map = get_format_map_from_video_format (format);
+
+  return map ? map->drm_fourcc : 0;
+}
+
 guint
 gst_va_chroma_from_video_format (GstVideoFormat format)
 {
@@ -312,14 +362,17 @@ gst_va_video_surface_format_from_image_format (GstVideoFormat image_format,
 }
 
 static GstVideoFormat
-find_gst_video_format_in_rgb32_map (VAImageFormat * image_format)
+find_gst_video_format_in_rgb32_map (VAImageFormat * image_format,
+    guint * drm_fourcc)
 {
   guint i, j;
 
   for (i = 0; i < G_N_ELEMENTS (rgb32_format_map); i++) {
     for (j = 0; j < G_N_ELEMENTS (rgb32_format_map[i].va_format); j++) {
-      if (va_format_is_same (&rgb32_format_map[i].va_format[j], image_format))
+      if (va_format_is_same (&rgb32_format_map[i].va_format[j], image_format)) {
+        *drm_fourcc = rgb32_format_map[i].drm_fourcc;
         return rgb32_format_map[i].format;
+      }
     }
   }
 
@@ -339,13 +392,14 @@ fix_map (gpointer data)
   GstVideoFormat format;
   VAImageFormat *image_format;
   struct FormatMap *map;
+  guint drm_fourcc = 0;
   guint i;
 
   for (i = 0; i < args->len; i++) {
     image_format = &args->image_formats[i];
     if (!va_format_is_rgb (image_format))
       continue;
-    format = find_gst_video_format_in_rgb32_map (image_format);
+    format = find_gst_video_format_in_rgb32_map (image_format, &drm_fourcc);
     if (format == GST_VIDEO_FORMAT_UNKNOWN)
       continue;
     map = get_format_map_from_video_format (format);
@@ -355,11 +409,14 @@ fix_map (gpointer data)
       continue;
 
     map->va_format = *image_format;
+    map->drm_fourcc = drm_fourcc;
 
-    GST_INFO ("GST_VIDEO_FORMAT_%s => { fourcc %"
-        GST_FOURCC_FORMAT ", %s, bpp %d, depth %d, R %#010x, G %#010x, "
-        "B %#010x, A %#010x }", gst_video_format_to_string (map->format),
+    GST_INFO ("GST_VIDEO_FORMAT_%s => { fourcc %" GST_FOURCC_FORMAT ", "
+        "drm fourcc %" GST_FOURCC_FORMAT ", %s, bpp %d, depth %d, "
+        "R %#010x, G %#010x, B %#010x, A %#010x }",
+        gst_video_format_to_string (map->format),
         GST_FOURCC_ARGS (map->va_format.fourcc),
+        GST_FOURCC_ARGS (map->drm_fourcc),
         (map->va_format.byte_order == 1) ? "LSB" : "MSB",
         map->va_format.bits_per_pixel, map->va_format.depth,
         map->va_format.red_mask, map->va_format.green_mask,
diff --git a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvavideoformat.h b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvavideoformat.h
index 609b27b597..9854c6bb05 100644
--- a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvavideoformat.h
+++ b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvavideoformat.h
@@ -33,6 +33,12 @@ GstVideoFormat        gst_va_video_format_from_va_fourcc  (guint fourcc);
 GST_VA_API
 guint                 gst_va_fourcc_from_video_format     (GstVideoFormat format);
 
+GST_VA_API
+GstVideoFormat        gst_va_video_format_from_drm_fourcc (guint fourcc);
+
+GST_VA_API
+guint                 gst_va_drm_fourcc_from_video_format (GstVideoFormat format);
+
 GST_VA_API
 guint                 gst_va_chroma_from_video_format     (GstVideoFormat format);
 
-- 
2.25.1

